src/sagemaker_xgboost_container/serving.py

```diff
@@ -65,22 +66,42 @@ def _set_mms_configs(is_multi_model, handler):
     Note: Ideally, instead of relying on env vars, this should be written directly to a config file.
     """
     max_content_length = os.getenv("MAX_CONTENT_LENGTH", DEFAULT_MAX_CONTENT_LEN)
+    max_workers = multiprocessing.cpu_count()
 
     if is_multi_model:
         os.environ["SAGEMAKER_NUM_MODEL_WORKERS"] = '1'
         os.environ["SAGEMAKER_MMS_MODEL_STORE"] = '/'
         os.environ["SAGEMAKER_MMS_LOAD_MODELS"] = ''
     else:
-        os.environ["SAGEMAKER_NUM_MODEL_WORKERS"] = str(multiprocessing.cpu_count())
+        os.environ["SAGEMAKER_NUM_MODEL_WORKERS"] = str(max_workers)
         os.environ["SAGEMAKER_MMS_MODEL_STORE"] = '/opt/ml/model'
         os.environ["SAGEMAKER_MMS_LOAD_MODELS"] = 'ALL'
 
     if not os.getenv("SAGEMAKER_BIND_TO_PORT", None):
         os.environ["SAGEMAKER_BIND_TO_PORT"] = str(PORT)
 
+    # Max heap size = max workers * max payload size * 1.2 (20% buffer) + 128 (base amount)
+    max_heap_size = ceil(max_workers * max_content_length * 1.2) + 128
+    os.environ["SAGEMAKER_MAX_HEAP_SIZE"] = str(max_heap_size) + 'm'
+    os.environ["SAGEMAKER_MAX_DIRECT_MEMORY_SIZE"] = os.environ["SAGEMAKER_MAX_HEAP_SIZE"]
+
     os.environ["SAGEMAKER_MAX_REQUEST_SIZE"] = str(max_content_length)
     os.environ["SAGEMAKER_MMS_DEFAULT_HANDLER"] = handler
 
+    # TODO: Revert config.properties.tmp to config.properties and add back in vmargs
+    # set with environment variables after MMS implements parsing environment variables
+    # for vmargs, update MMS section of final/Dockerfile.cpu to match, and remove the
+    # following code.
+    try:
+        with open('/home/model-server/config.properties.tmp', 'r') as f:
+            with open('/home/model-server/config.properties', 'w+') as g:
+                g.write("vmargs=-XX:-UseLargePages -XX:+UseG1GC -XX:MaxMetaspaceSize=32M -XX:+ExitOnOutOfMemoryError "
+                        + "-Xmx" + os.environ["SAGEMAKER_MAX_HEAP_SIZE"]
+                        + " -XX:MaxDirectMemorySize=" + os.environ["SAGEMAKER_MAX_DIRECT_MEMORY_SIZE"] + "\n")
+                g.write(f.read())
+    except Exception:
+        pass
```

Comment:
Is the server going to start without `config.properties`? If not, is it ok to fail silently here?

Reply:
I added this try-except because the path to the config.properties only exists on the actual container, which was causing the unit tests to fail. Is there a better way to do this?

Reply:
I just read the code comments above the try-except block. If the plan is to remove this code in the near future, then I guess it's okay. Reading `try: something, except Exception: pass` made me think "it's okay to fail this block".

Reply:
Yeah, ideally we want to have the config.properties as it was before and use the environment variables to get the vmargs. The issue is that the MMS team hasn't implemented environment variable parsing for vmargs, so this is a temporary solution until they release that feature.

Reply:
Minor: instead of concat'ing strings, can you use format and use variable names to put values in the string for readability?

Also, I'm ok with this temporary solution but another solution worth thinking about would've been to take files and a dict and do overrides to create a new config file and its location.